crates/primitives/transactions/src/compute_hash.rs

```diff
@@ -1,37 +1,85 @@
 use alloc::vec::Vec;
 
+use blockifier::transaction::transactions::L1HandlerTransaction as L1HandlerTransactionBlockifier;
 use mp_felt::Felt252Wrapper;
 use mp_hashers::HasherT;
+use starknet_api::data_availability::DataAvailabilityMode;
+use starknet_api::transaction::{
+    Calldata, DeclareTransaction, DeclareTransactionV0V1, DeclareTransactionV2, DeclareTransactionV3,
+    DeployAccountTransaction, DeployAccountTransactionV1, DeployAccountTransactionV3, DeployTransaction,
+    InvokeTransaction, InvokeTransactionV0, InvokeTransactionV1, InvokeTransactionV3, L1HandlerTransaction, Resource,
+    ResourceBoundsMapping, Transaction, TransactionHash,
+};
 use starknet_core::crypto::compute_hash_on_elements;
 use starknet_core::utils::starknet_keccak;
 use starknet_crypto::FieldElement;
 
-use super::{
-    DeclareTransaction, DeclareTransactionV0, DeclareTransactionV1, DeclareTransactionV2, DeployAccountTransaction,
-    HandleL1MessageTransaction, InvokeTransaction, InvokeTransactionV0, InvokeTransactionV1, Transaction,
-    UserTransaction, SIMULATE_TX_VERSION_OFFSET,
-};
-use crate::{DeployTransaction, UserOrL1HandlerTransaction, LEGACY_BLOCK_NUMBER, LEGACY_L1_HANDLER_BLOCK};
+use super::SIMULATE_TX_VERSION_OFFSET;
+use crate::LEGACY_BLOCK_NUMBER;
 
 const DECLARE_PREFIX: &[u8] = b"declare";
 const DEPLOY_ACCOUNT_PREFIX: &[u8] = b"deploy_account";
 const DEPLOY_PREFIX: &[u8] = b"deploy";
 const INVOKE_PREFIX: &[u8] = b"invoke";
 const L1_HANDLER_PREFIX: &[u8] = b"l1_handler";
+const L1_GAS: &[u8] = b"L1_GAS";
+const L2_GAS: &[u8] = b"L2_GAS";
 
 pub trait ComputeTransactionHash {
     fn compute_hash<H: HasherT>(
         &self,
         chain_id: Felt252Wrapper,
         offset_version: bool,
         block_number: Option<u64>,
-    ) -> Felt252Wrapper;
+    ) -> TransactionHash;
+}
+
+fn convert_calldata(calldata: Calldata) -> Vec<FieldElement> {
+    calldata.0.iter().map(|f| Felt252Wrapper::from(*f).into()).collect()
 }
 
-fn convert_calldata(data: &[Felt252Wrapper]) -> &[FieldElement] {
-    // Non-copy but less dangerous than transmute
-    // https://doc.rust-lang.org/std/mem/fn.transmute.html#alternatives
-    unsafe { core::slice::from_raw_parts(data.as_ptr() as *const FieldElement, data.len()) }
+fn prepare_resource_bound_value(resource_bounds_mapping: &ResourceBoundsMapping, resource: Resource) -> FieldElement {
```

Comment:
This function seems a bit esoteric to me, consider adding some clarifying comments.